doc/tuned.8
.TH "tuned" "8" "25 Feb 2009" "Phil Knirsch" "Adaptive system tuning daemon"
.SH NAME
tuned - dynamic adaptive system tuning daemon
.SH SYNOPSIS
\fBtuned\fP [\fIoptions\fP]
.SH DESCRIPTION
\fBtuned\fR is a dynamic adaptive system tuning daemon
that tunes system settings dynamically depending on
usage. For each hardware subsystem a specific monitoring plugin collects data
periodically. This information is then used by tuning plugins to change system
settings to lower or higher power saving modes in order to adapt to the current
usage. Currently monitoring and tuning plugins for CPU, ethernet network and ATA
harddisk devices are implemented.

In order to make it easier to switch between settings for different setups or
locations we now have an additional tool to switch between profiles.
Several pre defined profiles have already been included, but new ones can easily
be added or existing ones modified. For more details look at tuned-adm(1).
.SH OPTIONS
.TP 12
.BI \-d "\fR, \fP" \-\-daemon
This options starts tuned as a daemon as opposed to
in the foreground without forking at startup.
.TP 12
.BI \-c "\fR, \fP" \-\-conffile=" conffile"
Specify the name and path of the configuration file, default is \fB/etc/tuned.conf\fR.
.TP 12
.BI \-D "\fR, \fP" \-\-debug
Sets the highest logging level. This could be very useful when having trouble with tuned.
.SH "FILES"
.nf
/etc/tuned.conf
/usr/share/doc/tuned-*/README
.SH "SEE ALSO"
.LP
tuned.conf(5)
tuned\-adm(1)
.SH AUTHOR
Written by Phil Knirsch <pknirsch@redhat.com>.
.SH REPORTING BUGS
Report bugs to <pknirsch@redhat.com>. 


ChangeLog
[33mcommit adf6d855789f1e5738ae0a113a96d914b480e23f[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Tue Jan 29 16:05:50 2013 +0100

    udev rules: ask tuned to restart ktune instead of restarting it directly

[33mcommit afd26f62774daf81a9f95d256d415547300f7757[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Tue Jan 29 13:26:25 2013 +0100

    tuned: restart ktune 10 seconds after last SIGUSR2 was received
    
    When multiple devices are added into the system, udev restarted ktune
    daemon for each new device. This can cause races within ktune. This
    mechanism allows triggering just one ktune restart when multiple
    devices are inserted.

[33mcommit 6fc759a9fff5801a9f80e672207fc41f95b3ae7b[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Tue Jan 29 14:32:30 2013 +0100

    functions: implement support for readahead restore

[33mcommit f46c9d05477f93f31e3c5ee1c24637f4adde743c[m
Author: Jaroslav Škarvada <jskarvad@redhat.com>
Date:   Thu Jan 3 17:03:43 2013 +0100

    profiles: removed the nobarrier from the virtual-guest profile
    
    resolves: #886956

[33mcommit 355fd87b1b2304863fe04266e290dc99394587ad[m
Author: Jaroslav Škarvada <jskarvad@redhat.com>
Date:   Thu Jan 3 17:01:48 2013 +0100

    disabled transparent hugepages in latency-performance profile
    
    resolves: rhbz#887355

[33mcommit 66d8e396db3845463abc181c541ee09c686b1ff1[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Fri Dec 7 14:06:09 2012 +0100

    {disk,net}devstat: fix typo in usage

[33mcommit c1b7f35546a23b409ca0e1479d928b8d7f57c06b[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Oct 10 10:47:26 2012 +0200

    udev: add rule to restart ktune in new block device is added
    
    Resolves: #847445

[33mcommit 64b45e8a04b06c0f554c64b458c590c9904f7f5f[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Thu Sep 27 13:40:38 2012 +0200

    profiles: do not disable barriers on devices with write back cache
    
    Resolves: #801561

[33mcommit 9e8f67009568adce6b9a2f1d4c4c828b2192cba8[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Mon Sep 24 15:08:19 2012 +0200

    tuned: create PID file with proper permissions
    
    Resolves: #845336

[33mcommit 84536e5a5917282821a654cb5d47ac3505d9506e[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Mon Sep 24 14:51:59 2012 +0200

    ktune: validate value of ELEVATOR_TUNE_DEVS
    
    Resolves: #784308

[33mcommit d4c0b5a2d145da188b995419764f9c969b9afdbb[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Tue Mar 20 14:29:04 2012 +0100

    update systemd unit file install path
    
    required by /lib -> /usr/lib feature

[33mcommit 9dfa60a1f4166aab80e400d88a3ee28bcc256efc[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Tue Mar 20 14:16:42 2012 +0100

    spec: missing system-units build requirement

[33mcommit cbb2bf6e85117a0f245de5010e70c75a8ccc35b6[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Tue Mar 20 13:55:48 2012 +0100

    version bump

[33mcommit 0427cea0d801b03ab5fbfdeb3c26922c6eaa2247[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Mon Mar 5 08:51:53 2012 +0100

    netdevstat: cmdline error handling

[33mcommit 81590cf7705e982c7880c69d01ea6d02a6e7a2e4[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Mon Mar 5 08:45:50 2012 +0100

    diskdevstat: cmdline error handling

[33mcommit 3b10ac690bc6164b721236702cb5eb8493278f1e[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Fri Mar 2 09:59:29 2012 +0100

    virtual-guest: cleanup ktune.sh
    
    References: rhbz#740976, rhbz#740977

[33mcommit 90afd92add2facb02596a7e8e1341f252ae3684d[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Fri Mar 2 09:00:26 2012 +0100

    virtual-host: tune virtual memory settings
    
    References: rhbz#740976

[33mcommit 78012939d465405587a6607838f15e8c06cbd5a3[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Jan 11 13:38:28 2012 +0100

    tuned-adm: sort list of available profiles

[33mcommit e30a77ede22df20232547b6243107b57ed0ad0ee[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Jan 11 13:35:45 2012 +0100

    tuned-adm man: add virtual-host profile decription

[33mcommit 5984ad2de7e2e5ee76d54191f03515ae55739a4e[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Jan 11 13:26:31 2012 +0100

    add virtual-host profile

[33mcommit 0e4826f78b14f219c876bc9d3b35834cd7a18311[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Jan 11 12:59:48 2012 +0100

    tuned-adm: fix crash when active-profile file is missing

[33mcommit 0754d462ff72ed69b704b3a4a4d7a5ecc77df655[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Jan 11 12:55:22 2012 +0100

    tuned-adm man: add virtual-guest profile decription

[33mcommit 8bb2e57c523cc27915b422d077ba5c644bd85f6a[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Jan 11 12:48:19 2012 +0100

    add virtual-guest profile

[33mcommit 8bd483a47c4e00cf106fb16d50ac5d0c61f21ab8[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Jan 11 12:40:23 2012 +0100

    functions: support for disk scheduler quantum setting

[33mcommit b5297bcc8cdb71b7c7fff17b7a18ea194952cc1c[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Jan 11 12:39:47 2012 +0100

    functions: update disks detection mechanism

[33mcommit f11bae54bc26e2969ab1413995baeae472e75b60[m
Author: Jaroslav Škarvada <jskarvad@redhat.com>
Date:   Sat Dec 3 15:31:29 2011 +0100

    set_radeon_powersave: fixed profile name

[33mcommit 229b5f7874c406f702b7f1fd9d4e18a1df0e575e[m
Author: Jaroslav Škarvada <jskarvad@redhat.com>
Date:   Sat Dec 3 15:14:06 2011 +0100

    Changed Radeon PM to auto in laptop-battery-powersave
    
    It should be safer for presenters, because the more
    aggresive dynpm doesn't support multi heads.

[33mcommit 60f78e3dc83907d558913473acc145ce864d4c26[m
Author: Jaroslav Škarvada <jskarvad@redhat.com>
Date:   Sat Dec 3 14:32:39 2011 +0100

    Improved radeon powersavings settings in tuned profiles

[33mcommit 9280e4ee158792973112ea78c3848bc25609c51e[m

tuned.py
#!/usr/bin/python -Es
#
# tuned: daemon for monitoring and adaptive tuning of system devices
#
# Copyright (C) 2008-2013 Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
#

import argparse
import os
import sys
import traceback
import tuned.logs
import tuned.daemon
import tuned.exceptions
import tuned.consts as consts
import tuned.version as ver

def error(message):
	print >>sys.stderr, message

if __name__ == "__main__":
	parser = argparse.ArgumentParser(description = "Daemon for monitoring and adaptive tuning of system devices.")
	parser.add_argument("--daemon", "-d", action = "store_true", help = "run on background")
	parser.add_argument("--debug", "-D", action = "store_true", help = "show/log debugging messages")
	parser.add_argument("--log", "-l", nargs = "?", const = consts.LOG_FILE, help = "log to file, default file: " + consts.LOG_FILE)
	parser.add_argument("--pid", "-P", nargs = "?", const = consts.PID_FILE, help = "write PID file, default file: " + consts.PID_FILE)
	parser.add_argument("--no-dbus", action = "store_true", help = "do not attach to DBus")
	parser.add_argument("--profile", "-p", action = "store", type=str, metavar = "name", help = "tuning profile to be activated")
	parser.add_argument('--version', "-v", action = "version", version = "%%(prog)s %s.%s.%s" % (ver.TUNED_VERSION_MAJOR, ver.TUNED_VERSION_MINOR, ver.TUNED_VERSION_PATCH))
	args = parser.parse_args(sys.argv[1:])

	if os.geteuid() != 0:
		error("Superuser permissions are required to run the daemon.")
		sys.exit(1)

	log = tuned.logs.get()
	if args.debug:
		log.setLevel("DEBUG")

	try:
		if args.daemon:
			if args.log is None:
				args.log = consts.LOG_FILE
			log.switch_to_file(args.log)
		else:
			if args.log is not None:
				log.switch_to_file(args.log)

		app = tuned.daemon.Application(args.profile)

		if not args.no_dbus:
			app.attach_to_dbus(consts.DBUS_BUS, consts.DBUS_OBJECT, consts.DBUS_INTERFACE)

		# always write PID file
		if args.pid is None:
			args.pid = consts.PID_FILE

		if args.daemon:
			app.daemonize(args.pid)
		else:
			app.write_pid_file(args.pid)
		app.run()

	except tuned.exceptions.TunedException as exception:
		if (args.debug):
			traceback.print_exc()
		else:
			error(str(exception))
			sys.exit(1)
